native-addons-are-close-to-being-on-par-with-js-modules.md

```diff
@@ -0,0 +1,113 @@
+
+```
+Error: The module '/home/user/node_hello_world/build/Release/hello.node'
+was compiled against a different Node.js version using
+NODE_MODULE_VERSION 57. This version of Node.js requires
+NODE_MODULE_VERSION 64. Please try re-compiling or re-installing
+the module (for instance, using `npm rebuild` or `npm install`).
+    at Object.Module._extensions..node (internal/modules/cjs/loader.js:718:18)
+    at Module.load (internal/modules/cjs/loader.js:599:32)
+    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
+    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
+    at Module.require (internal/modules/cjs/loader.js:637:17)
+    at require (internal/modules/cjs/helpers.js:22:18)
+    at bindings (/home/user/node_hello_world/node_modules/bindings/bindings.js:76:44)
+    at Object.<anonymous> (/home/user/node_hello_world/hello.js:1:94)
+    at Module._compile (internal/modules/cjs/loader.js:689:30)
+    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
+```
+
+Such errors can also be avoided by the use of prebuild utilities, however, it places even more burden on addon maintainers, namely, to rebuild the binaries they provide at every release of Node.js. It also prevents application maintainers from simply dropping a new version of Node.js into their production environment. Indeed they need to reinstall their application even though no part of its functionality has changed.
+
+[N-API][] was introduced to address the problem of the binary compatibility of native addons with versions of Node.js against which they were build *as well as all subsequent versions*. It abstracts away essential JavaScript engine and Node.js APIs, presenting them to native addon maintainers as a collection of ABI-stable APIs. The stability of N-API is assured by the fact that all APIs are C APIs rather than C++ APIs and by a commitment on the part of the Node.js core developer community to avoid modifying the interface in an incompatible way.
+
```

Comment:
I think we should move some of this up front or add a caveat up front that a number of the problems you are going to cover have been addressed by N-API.  If you stop reading before this point you might believe these are still unsolved problems.

Reply:
I think we also want to expand this. N-API was meant to help with more that ABI stability

* avoid the necessity to update module due to changes in v8, even if you recompile on installation (the breakage of modules on 12.x illustrates where this would help)
* Allow binaries to run with multiple versions avoiding need to recompile
* Allow innovation at the JavaScript runtime layer by allowing addons to run with different Engines.  This allows different implementations of node for specific situations (for example resource constrained environments)

Reply:
@mhdawson I've added a blurb at the beginning to immediately indicate that the article intends to focus on newly developed solutions. I've also expanded on the possible uses of N-API.